scertec.go

```diff
@@ -0,0 +1,189 @@
+// There is one parsedCert per secret name but its latest
+// pointer gets updated whenever the PEM from the secret changes.
+type parsedCert struct {
+	latest atomic.Pointer[pemAndParsed]
+}
+
+// pemAndParsed are the PEM and parsed certificate for a particular
+// version of the secret. (We don't know the secret version but
+// we notice when the PEM bytes change). The pem and parsed values
+// are immutable once set.
+type pemAndParsed struct {
+	pem    []byte
+	parsed *tls.Certificate
+	hits   atomic.Int64 // for tests
+}
+
+func secretName(devMode bool, domain string, certType string) string {
+	if devMode {
+		return "dev/certd/https-cert/" + domain + "/" + certType
+	}
+	return "prod/certd/https-cert/" + domain + "/" + certType
+}
+
+// NewClient returns a new HTTPS cert client. It blocks until all the needed
+// secrets are available for retrieval by the Secret method, or ctx ends.  The
+// context passed to NewStore is only used for initializing the store.
+//
+// As a special case for testing, if domains contains "DEV" as the last element,
+// the Client will use "dev/" secrets instead of "prod/" secrets.
+func NewClient(ctx context.Context, c setec.Client, cache setec.Cache, domains ...string) (*Client, error) {
+	var secretNames []string
+	devMode := len(domains) > 0 && domains[len(domains)-1] == "DEV"
+	if devMode {
+		domains = domains[:len(domains)-1]
+	}
+	for _, d := range domains {
+		secretNames = append(secretNames,
+			secretName(devMode, d, "rsa"),
+			secretName(devMode, d, "ecdsa"),
+		)
+	}
+	st, err := setec.NewStore(ctx, setec.StoreConfig{
+		Client:  c,
+		Secrets: secretNames,
+		Cache:   cache,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		sc:      c,
+		st:      st,
+		devMode: devMode,
+	}, nil
+}
+
+// GetCertificate returns the RSA or ECDSA certificate for hello.ServerName.
+//
+// It is the signature needed by tls.Config.GetCertificate.
+func (c *Client) GetCertificate(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	typ := "rsa"
+	if supportsECDSA(hello) && !c.forceRSA.Load() {
+		typ = "ecdsa"
+	}
+	secName := secretName(c.devMode, hello.ServerName, typ)
+	sec := c.st.Secret(secName)
+	if sec == nil {
+		return nil, errors.New("invalid server name")
+	}
+	return c.parsedCert(secName, sec.Get())
+}
+
+func (c *Client) parsedCert(secName string, pems []byte) (*tls.Certificate, error) {
+	pci, ok := c.parsed.Load(secName)
+	if !ok {
+		pci, _ = c.parsed.LoadOrStore(secName, &parsedCert{})
+	}
+	pc := pci.(*parsedCert)
+
+	latest := pc.latest.Load()
+	if latest != nil && bytes.Equal(latest.pem, pems) {
+		// Common case; the cert hasn't changed.
+		latest.hits.Add(1)
+		return latest.parsed, nil
+	}
+
+	b, certPEMBlock := pem.Decode(pems)
+	if b == nil {
+		return nil, errors.New("invalid PEM")
+	}
+	keyPEMSize := len(pems) - len(certPEMBlock)
+	keyPEMBlock := pems[:keyPEMSize]
+
+	cert, err := tls.X509KeyPair(certPEMBlock, keyPEMBlock)
```

Comment:
eh, I dunno? What's the worst that can happen, we serve a different type TLS handshake error here? bad cert vs no cert? It's not like we'd leak a private key.

Reply:
One possibility is scertecd writing a new cert to the wrong path. Without a check here, that would serve the wrong cert to clients and they would reject it. With a check, we'd have some time to notice that a new cert is not getting picked up because of wrong CN